### PR TITLE
refactor(Make all links readable in Voucher)

### PR DESCRIPTION
**Pull request by lomamech**
- Update search modal, make select User in readable format
- Update API for finance/voucher fixe the filter in select multiple
  transaction type
- Update select transanction type in readable format

this issue reference #2139 and #2130

### CODE REVIEW

File: server/controllers/finance/vouchers.js [hunk 1]
```diff
@@ -115,7 +115,7 @@ function find(options) {
   let typeIds = [];
 
   if (options.type_ids) {
-    typeIds = typeIds.concat(options.type_ids);
+    typeIds = options.type_ids.toString();
```
Comment (jniles) [inline server/controllers/finance/vouchers.js]: Why was this change necessary?  The previous version escaped an array of values, but this one turns them into a string without escaping.  In theory, I could send `?type_ids=1 OR 1=1` and bypass all filtering via an SQL injection.

File: client/src/modules/vouchers/voucher-registry.ctrl.js [hunk 2]
```diff
@@ -116,6 +116,9 @@ function VoucherController(
   function search() {
     var filtersSnapshot = Vouchers.filters.formatHTTP();
 
+    // Keep the last filters made during the last search and keeps the display in a readable format
```
Comment (jniles) [inline client/src/modules/vouchers/voucher-registry.ctrl.js]: This is an interesting way to pass filters in ... but it would probably be better to use the service to get the filters in the modal controller.  It should be as simple as importing `VoucherService` into the `ModalController` and then doing `var filters = VoucherService.filters.formatView();`

File: client/src/modules/vouchers/modals/search.modal.js [hunk 3]
```diff
@@ -74,8 +98,21 @@ function VoucherRegistrySearchModalController(ModalInstance, filters, Notify, mo
 
     // push all searchQuery values into the changes array to be applied
     angular.forEach(vm.searchQueries, function (value, key) {
-      if (angular.isDefined(value)) {
-        changes.post({ key : key, value : value });
+      if (angular.isDefined(value)) {
```
Comment (jniles) [inline client/src/modules/vouchers/modals/search.modal.js]: This is pretty complicated code.

Why don't you track display values in a separate object.  It doesn't even need to be tied to the view - something like this:
```js
// searchQueries is the same id:value pair
vm.searchQueries = {};
// displayValues will be an id:displayValue pair
var displayValues = {};
```
... then, each individual method can define how to display it's display value.  For example, for `user_id`, you can just do this:
```js
// search on user.id, but show user.display_name in the filters
vm.onSelectUser = function onSelectUser(user) {
  vm.searchQueries.user_id = user.id;
  displayValues.user_id = user.display_name;
};
```
... and then in the submit method ...
```js
angular.forEach(vm.searchQueries, function (value, key) {
  if (angular.isDefined(value)) {
    // default to the original value if no display value is defined
    var displayValue = displayValues[key] || value;
    changes.post({ key: key, value: value, displayValue: displayValue });
   }
});
```

File: client/src/modules/vouchers/voucher-registry.ctrl.js (hunk 2 above)
Comment (jniles) [reply client/src/modules/vouchers/voucher-registry.ctrl.js]: Having read all the code now, I actually don't think you need to share the view values at all.  It should all be driven by the search modal.

**Review by jniles — CHANGES_REQUESTED** [review]
@lomamech this is on the right track, but there is a much easier way to handle these changes.  I've proposed it in the comments below.

File: server/controllers/finance/vouchers.js (hunk 1 above)
Comment (lomamech) [reply server/controllers/finance/vouchers.js]: Because with typeIds = typeIds.concat (options.type_ids); the query filters only the first selected element
error

Comment (jniles) [reply server/controllers/finance/vouchers.js]: Try using ES6 then - typeIds = [...options.type_ids];

Chances are `type_ids` is no longer being sent as an array, which is what the server API expects.

File: client/src/modules/vouchers/modals/search.modal.js [hunk 4]
```diff
@@ -13,32 +14,67 @@ VoucherRegistrySearchModalController.$inject = [
...
+
   if (filters.limit) {
     vm.defaultQueries.limit = filters.limit;
   }
 
   vm.onTransactionTypesChange = function onTransactionTypesChange(transactionTypes) {
     vm.searchQueries.type_ids = transactionTypes;
+    var typeText = '/';
+
+    transactionTypes.forEach(function (typeId) {
```
Comment (jniles) [inline client/src/modules/vouchers/modals/search.modal.js:63]: This solution is okay for now.

What we should probably do is modify the `onTransactionTypeChange()` callback to send back the list of transaction types, rather than a list of IDs.  That way we can simply use their labels instead of downloading transaction types twice.

File: server/controllers/finance/vouchers.js [hunk 5]
```diff
@@ -144,7 +144,7 @@ function find(options) {
   // @todo - could this be improved
   filters.custom('entity_uuid', 'v.uuid IN (SELECT DISTINCT voucher_uuid FROM voucher_item WHERE entity_uuid = ?)');
 
-  filters.custom('type_ids', 'v.type_id IN (?)', typeIds);
+  filters.custom('type_ids', 'v.type_id IN (?)', [typeIds]);
```
Comment (jniles) [inline server/controllers/finance/vouchers.js:147]: 💃

**Review by jniles — APPROVED** [review]
Looks good to me!

bors r+
